hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerStateMachine.java

```diff
@@ -349,22 +357,51 @@ private Message runCommand(OMRequest request, long trxLogIndex) {
   /**
    * Update lastAppliedIndex term and it's corresponding term in the
    * stateMachine.
+   * @param flushedEpochs
+   */
+  public void updateLastAppliedIndex(List<Long> flushedEpochs) {
+    Preconditions.checkArgument(flushedEpochs.size() > 0);
+    updateLastAppliedIndex(flushedEpochs.get(flushedEpochs.size() -1),
+        -1L, flushedEpochs, true);
+  }
+
+  /**
+   * Update State machine lastAppliedTermIndex.
    * @param lastFlushedIndex
+   * @param currentTerm
+   * @param flushedEpochs - list of ratis transactions flushed to DB. If it
+   * is just one index and term, this can be set to null.
+   * @param checkMap - if true check applyTransactionMap, ratisTransaction
+   * Map and update lastAppliedTermIndex accordingly, else check
+   * lastAppliedTermIndex and update it.
    */
-  public synchronized void updateLastAppliedIndex(long lastFlushedIndex) {
-    Long appliedTerm = null;
-    long appliedIndex = -1;
-    for(long i = getLastAppliedTermIndex().getIndex() + 1;
-        i <= lastFlushedIndex; i++) {
-      final Long removed = applyTransactionMap.remove(i);
-      if (removed == null) {
-        break;
+  private synchronized void updateLastAppliedIndex(long lastFlushedIndex,
+      long currentTerm, List<Long> flushedEpochs, boolean checkMap) {
+    if (checkMap) {
+      Long appliedTerm = null;
+      long appliedIndex = -1;
+      for (long i = getLastAppliedTermIndex().getIndex() + 1; ; i++) {
+        if (flushedEpochs.contains(i)) {
+          appliedIndex = i;
+          final Long removed = applyTransactionMap.remove(i);
+          appliedTerm = removed;
+        } else if (ratisTransactionMap.containsKey(i)) {
+          final Long removed = ratisTransactionMap.remove(i);
+          appliedTerm = removed;
+          appliedIndex = i;
+        } else {
+          break;
+        }
```

Comment:
@bharatviswa504, I am trying to think of cases in which this approach might fail. This scenario might not be possible so correct me if I am wrong.
Lets say we have multiple executors and the DoubleBuffer flushes the transactions as below.
Flush 1 -> Txns 1, 2, 4
Flush 2 -> Txns 3, 5, 6
Flush 3 -> Txns 7, 8, 9, 10

Now, when the 1st flush happens and it calls updateLaspAppliedIndex, for txn 1 and 2 there will not be a problem. Then it will look for txn 3 which is not a part of either flushedEpochs or ratisTransactionMap. Hence, the loop will break and the SM lastAppliedIndex will be updated to 2.
The next flush has txns 3, 5 and 6. The same scenario will be repeated and the SM lastAppliedIndex will be updated to 3.
Now when the 3rd flush calls updateLastAppliedIndex, the SM lastAppliedIndex will not be updated as the we have no record of txns 4, 5 and 6 being flushed to disk.

Reply:
I think with multiple executors this can certainly possible. Thanks for catching it. Fixed to handle this issue in latest commits.